### PR TITLE
Support `lr0=auto` to find the optimal LR for fine-tuning

### PR DESCRIPTION
**Pull request by Y-T-G**
<table>
<tr><th>Example Objects365 checkpoint fine-tuned on COCO</th></tr>
<tr><td>image</td></tr>
<tr><td>lr_curve_probe_w0 125</td></tr>
</table>

A more dynamic algorithm to select the best LR for training. LR significantly affects fine-tuning performance. The default values currently Ultralytics uses are designed for scratch training. It is too aggressive for fine-tuning or starting from pretrained checkpoint, which is what the vast majority of the users do.

This PR adds an LR sweep based on the idea proposed in [Cyclical Learning Rates for Training Neural Networks](https://arxiv.org/abs/1506.01186) and improved on by [AdaLRS](https://arxiv.org/abs/2506.13274) (NeurIPS 2025).

The current LR fitting algorithm doesn't take into account the model and the dataset. While the PR uses a learning rate range test on the current dataset using the model in question, and uses it to fit `lr0` and `warmup_bias_lr`, scaling `lrf` to keep the final rate usable, at the cost of training start time.

It only runs if `lr0=auto`, with any optimizer including `optimizer=auto`.

Training from scratch still uses the current defaults because those work best for scratch training.

**Fine-tuning result**, `yolo26s` Objects365 checkpoint fine-tuned on COCO, 70 epochs, batch 128, `lrf=0.88219`, identical augmentation in both runs. This is the pair plotted above.

| run | `lr0` | `warmup_bias_lr` | `momentum` | mAP50-95 | mAP50-95 o2m |
| --- | --- | --- | --- | --- | --- |
| `optimizer=MuSGD lr0=auto` | 0.00025 | 0.00025 | 0.937 | **47.80** | **48.64** |
| current behaviour | 0.01 | 0.0 | 0.9 | 44.21 | 45.27 |

The fitted run gains **3.59 mAP50-95**. The second row spells out what `optimizer=auto` resolves to today, so it runs the pre-PR defaults. Those defaults start 40 times hotter than the fitted rate, and the run never catches up.

**Limits**

- The fitted `lr0` is capped at 0.01 to avoid NaN loss.
- Sweeps that do not bracket an optimum, shorter runs with insufficient steps, and `resume` all fall back to the equation that was already in use before.
- Cannot be made default, because it can cost around a minute to run the range test before the training starts

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary

Added `lr0="auto"` fine-tuning support that runs a dataset-specific learning-rate range test and applies fitted `lr0`, `warmup_bias_lr`, and, when appropriate, `lrf` values before training.

### 📊 Key Changes

- Added `Trainer._find_lr()` to sweep learning rates, restore the model and optimizer state, detect a suitable rate, and rebuild the scheduler.
- Enabled `lr0="auto"` for all optimizers, including `optimizer=auto`; an initial equation-based rate is used if the sweep cannot produce a valid fit.
- Restricted automatic fitting to fine-tuning runs; scratch training retains the existing default learning-rate behavior.
- Added fallback handling for short runs, unbracketed sweeps, unresolved descent peaks, runtime errors, and resumed training.
- Updated configuration validation, default settings, training argument documentation, and fine-tuning guidance with Python and CLI examples.

### 🎯 Purpose & Impact

- Fine-tuning can now select learning-rate settings from the current model and dataset instead of relying solely on fixed heuristics.
- `lr0="auto"` adds a range-test delay before training, and explicit `lrf` values are preserved while the default `lrf` may be adjusted to maintain a usable final learning rate.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/engine/trainer.py [hunk 1]
```diff
@@ -275,6 +275,114 @@ def _setup_ddp(self):
...
+                if self.world_size > 1:
+                    raise  # a rank that skips its backward would desynchronize the others
+                LOGGER.warning(f"{colorstr('LR finder:')} sweep stopped early, {e}")
+                self._clear_memory()
+                break
+            self.optimizer_step()
+            if self.world_size > 1:
+                dist.all_reduce(total)  # every rank fits the same global curve, so no result broadcast is needed
+            losses.append(total.item() / (self.accumulate * max(self.world_size, 1)))
+            if not math.isfinite(losses[-1]) or losses[-1] > 4 * min(losses):
```
Comment (UltralyticsAssistant) [inline ultralytics/engine/trainer.py:341]: ⚠️ **HIGH**: When a sweep loss becomes NaN after enough points, this branch stops but leaves the NaN in `losses`. The centered convolution and `argmin` can then select a NaN window, causing `polyfit` to produce NaN and line 375 to write `nan` into `lr0` and the optimizer groups instead of using the documented equation fallback. Exclude non-finite samples before fitting or fall back whenever the recorded curve contains one.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed configuration validation, trainer setup, LR-finder state restoration, optimizer integration, and the training-loop refactor. The implementation is generally coherent, but the non-finite-loss path can propagate NaN into the fitted learning rate and corrupt the subsequent training run.

<details><summary>💬 Posted 1 inline comment</summary>

- ⚠️ **HIGH** `ultralytics/engine/trainer.py:344` When a sweep loss becomes NaN after enough points, this branch stops but leaves the NaN in `losses`. The centered convolution and `argmin` can then select a NaN window, causing `polyfit` to produce NaN and line 375 to write `nan` into `lr0` and the optimizer groups instead of using the documented equation fallback. Exclude non-finite samples before fitting or fall back whenever the recorded curve contains one.
</details>

File: ultralytics/engine/trainer.py (hunk 1 above)
Comment (Y-T-G) [reply ultralytics/engine/trainer.py:341]: This cannot happen. The sweep breaks on the first non-finite value, so a NaN is always the last element of `losses` and never has a successor.

With `mode="valid"` and window 5, a final NaN lands in exactly one window, `y[-1]`. `np.argmin` returns that index, so `edge == len(y) - 1`, and the next line rejects it:

```python
if not 5 <= edge < len(y) - 1:
    LOGGER.warning(f"{colorstr('LR finder:')} sweep did not bracket an optimum, using the 'lr0' equation")
    return
```

Checked against a synthetic curve that falls, turns up, then ends non-finite:

| final sample | `len(y)` | NaN windows | `argmin` | passes guard |
| --- | --- | --- | --- | --- |
| `nan` | 62 | 1 | 61 | no |
| `inf` | 62 | 0 | 47 | yes |
| none | 61 | 0 | 47 | yes |

`inf` produces no NaN windows and never wins an `argmin`, so it is ignored as a sample and the real minimum is still found. `polyfit` is never reached with a non-finite input in either case.

**Comment by Y-T-G** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-16 17:17:29 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/31961104804): `ci.yml` run 29326
